File: 07_linked_list/07_12_add_two_numbers/add_two_numbers.cpp
#include <iostream>

class ListNode
{
public:
    int val;
    ListNode *next;

    ListNode(int value = 0, ListNode *nextNode = nullptr) : val(value), next(nextNode) {}
};

ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
{
    ListNode *head = new ListNode();
    ListNode *trav = head;

    int carryOver = 0;

    while (l1 || l2 || carryOver == 1)
    {
        int sumVal = 0;

        if (l1)
        {
            sumVal += l1->val;
            l1 = l1->next;
        }

        if (l2)
        {
            sumVal += l2->val;
            l2 = l2->next;
        }

        sumVal += carryOver;
        carryOver = sumVal / 10;

        trav->next = new ListNode(sumVal % 10);
        trav = trav->next;
    }

    return head->next;
}

void printList(ListNode *head)
{
    ListNode *current = head;
    while (current)
    {
        std::cout << current->val << " -> ";
        current = current->next;
    }
    std::cout << "null";
}

void testAddition(const std::string &label, ListNode *l1, ListNode *l2)
{
    std::cout << label;
    printList(addTwoNumbers(l1, l2));
    std::cout << std::endl;
}

int main()
{
    // Test Case 1: Normal addition
    ListNode *l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
    ListNode *l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
    testAddition("Test Case 1:", l1, l2); // Output: 7 -> 0 -> 8

    // Test Case 2: Unequal lengths with carry-over
    ListNode *l3 = new ListNode(9, new ListNode(9, new ListNode(9)));
    ListNode *l4 = new ListNode(1);
    testAddition("Test Case 2:", l3, l4); // Output: 0 -> 0 -> 0 -> 1

    // Test Case 3: One list is empty
    ListNode *l5 = new ListNode(1, new ListNode(8, new ListNode(2)));
    ListNode *l6 = nullptr;
    testAddition("Test Case 3:", l5, l6); // Output: 1 -> 8 -> 2

    // Test Case 4: Both lists are empty
    ListNode *l7 = nullptr;
    ListNode *l8 = nullptr;
    testAddition("Test Case 4:", l7, l8); // Output: null

    // Remember to free allocated memory
    delete l1;
    delete l2;
    delete l3;
    delete l4;
    delete l5;

    return 0;
}

File: 07_linked_list/07_12_add_two_numbers/add_two_numbers.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(l1, l2):
    head = ListNode()
    trav = head

    carry_over = 0

    while l1 or l2 or carry_over == 1:
        sum_val = 0

        if l1:
            sum_val += l1.val
            l1 = l1.next

        if l2:
            sum_val += l2.val
            l2 = l2.next

        sum_val += carry_over
        carry_over = sum_val // 10

        trav.next = ListNode(sum_val % 10)
        trav = trav.next

    return head.next

def test_addition(label, l1, l2):
    print(label, end=" ")
    print_list(addTwoNumbers(l1, l2))
    print()

def print_list(head):
    current = head
    while current:
        print(current.val, "-> ", end="")
        current = current.next
    print("None")

# Test Case 1: Normal addition
l1 = ListNode(2, ListNode(4, ListNode(3)))
l2 = ListNode(5, ListNode(6, ListNode(4)))
test_addition("Test Case 1:", l1, l2)  # Output: 7 -> 0 -> 8

# Test Case 2: Unequal lengths with carry-over
l3 = ListNode(9, ListNode(9, ListNode(9)))
l4 = ListNode(1)
test_addition("Test Case 2:", l3, l4)  # Output: 0 -> 0 -> 0 -> 1

# Test Case 3: One list is empty
l5 = ListNode(1, ListNode(8, ListNode(2)))
l6 = None
test_addition("Test Case 3:", l5, l6)  # Output: 1 -> 8 -> 2

# Test Case 4: Both lists are empty
l7 = None
l8 = None
test_addition("Test Case 4:", l7, l8)  # Output: None
